File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/mouse/HotbarView.java
package net.kdt.pojavlaunch.customcontrols.mouse;

import android.content.Context;
import android.graphics.Color;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import androidx.annotation.Nullable;

import com.arata.anim.AnimPlayer;
import com.arata.anim.animations.Animations;
import com.arata.yukarilauncher.event.single.MCOptionChangeEvent;
import com.arata.yukarilauncher.event.single.RefreshHotbarEvent;
import com.arata.yukarilauncher.event.value.HotbarChangeEvent;
import com.arata.yukarilauncher.feature.MCOptions;
import com.arata.yukarilauncher.setting.AllSettings;
import com.arata.yukarilauncher.setting.AllStaticSettings;
import com.arata.yukarilauncher.ui.subassembly.hotbar.HotbarType;
import com.arata.yukarilauncher.ui.subassembly.hotbar.HotbarUtils;

import net.kdt.pojavlaunch.GrabListener;
import net.kdt.pojavlaunch.LwjglGlfwKeycode;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.utils.MathUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.lwjgl.glfw.CallbackBridge;

public class HotbarView extends View implements View.OnLayoutChangeListener, Runnable {
    private final TapDetector mDoubleTapDetector = new TapDetector(2, TapDetector.DETECTION_METHOD_DOWN);
    private View mParentView;
    private static final int[] HOTBAR_KEYS = {
            LwjglGlfwKeycode.GLFW_KEY_1, LwjglGlfwKeycode.GLFW_KEY_2,   LwjglGlfwKeycode.GLFW_KEY_3,
            LwjglGlfwKeycode.GLFW_KEY_4, LwjglGlfwKeycode.GLFW_KEY_5,   LwjglGlfwKeycode.GLFW_KEY_6,
            LwjglGlfwKeycode.GLFW_KEY_7, LwjglGlfwKeycode.GLFW_KEY_8, LwjglGlfwKeycode.GLFW_KEY_9};
    private final DropGesture mDropGesture = new DropGesture(new Handler(Looper.getMainLooper()));
    private final GrabListener mGrabListener = new GrabListener() {
/**
 * 「on Grab State」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
        @Override
        public void onGrabState(boolean isGrabbing) {
            mLastIndex = -1;
            mDropGesture.cancel();
        }
    };

    private int mWidth;
    private int mLastIndex = -1;
    private int mGuiScale;

    //调整判定框宽高时，用这个动画播放器播放一个淡化动画，来给用户一个当前判定框范围的反馈
    private final AnimPlayer adjustAnimPlayer = new AnimPlayer();
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public HotbarView(Context context) {
        super(context);
        init();
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public HotbarView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public HotbarView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */

    @SuppressWarnings("unused") // You suggested me this constructor, Android
    public HotbarView(Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }
/**
 * 「init」メソッド。
 * このクラスに定義された機能メソッドです。
 */
    private void init() {
        setAlpha(0);
        setBackgroundColor(Color.parseColor("#80E64242"));
        adjustAnimPlayer.duration(800);
        adjustAnimPlayer.apply(new AnimPlayer.Entry(this, Animations.FadeOut));
    }
/**
 * 「on Attached To Window」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        EventBus.getDefault().register(this);
        ViewParent parent = getParent();
        if(parent == null) return;
        if(parent instanceof View) {
            mParentView = (View) parent;
            mParentView.addOnLayoutChangeListener(this);
        }
        adaptiveReset();
        CallbackBridge.addGrabListener(mGrabListener);
    }
/**
 * 「on Detached From Window」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        CallbackBridge.removeGrabListener(mGrabListener);
        EventBus.getDefault().unregister(this);
    }

    /**
     * Hotbar更新イベントを受信したときに判定領域を更新します。
     * @param event 更新イベント
     */
    @Subscribe
    public void event(RefreshHotbarEvent event) {
        post(this);
    }

    /**
     * options.txtファイルが変更されたときに判定領域を更新します。
     * GUIサイズを確認する必要があります。
     * @param event 更新イベント
     */
    @Subscribe
    public void event(MCOptionChangeEvent event) {
        mGuiScale = MCOptions.INSTANCE.getMcScale();
        post(this);
    }

    /**
     * 判定枠の幅と高さを手動で調整するイベントを監視します。
     * @param event 変更イベント
     */
    @Subscribe
    public void event(HotbarChangeEvent event) {
        manualReset(event.getWidth(), event.getHeight(), true);
    }

    /**
     * マージンレイアウトパラメータを取得します。
     */
    private ViewGroup.MarginLayoutParams getMarginLayoutParams() {
        ViewGroup.LayoutParams layoutParams = getLayoutParams();
        if (!(layoutParams instanceof ViewGroup.MarginLayoutParams))
            throw new RuntimeException("Incorrect LayoutParams type, expected ViewGroup.MarginLayoutParams");
        return (ViewGroup.MarginLayoutParams) layoutParams;
    }
/**
 * 「adaptive Reset」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private void adaptiveReset() {
        ViewGroup.MarginLayoutParams marginLayoutParams = getMarginLayoutParams();
        int height;
        marginLayoutParams.width = mWidth = mcScale(180);
        marginLayoutParams.height = height = mcScale(20);
        marginLayoutParams.leftMargin = (CallbackBridge.physicalWidth / 2) - (mWidth / 2);
        marginLayoutParams.topMargin = CallbackBridge.physicalHeight - height;
        setLayoutParams(marginLayoutParams);
    }
/**
 * 「manual Reset」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private void manualReset(int width, int height, boolean playAnim) {
        ViewGroup.MarginLayoutParams marginLayoutParams = getMarginLayoutParams();
        marginLayoutParams.width = mWidth = width;
        marginLayoutParams.height = height;
        marginLayoutParams.leftMargin = Tools.currentDisplayMetrics.widthPixels / 2 - width / 2;
        marginLayoutParams.topMargin = Tools.currentDisplayMetrics.heightPixels - height;
        setLayoutParams(marginLayoutParams);
        if (playAnim) adjustAnimPlayer.start();
    }
/**
 * タッチイベントを処理します。
 * ユーザーからのタッチ入力を検出し、適切なアクションを実行します。
 */

    @SuppressWarnings("ClickableViewAccessibility") // performClick does not report coordinates.
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if(!CallbackBridge.isGrabbing()) return false;
        boolean hasDoubleTapped = mDoubleTapDetector.onTouchEvent(event);

        // Check if we need to cancel the drop event
        int actionMasked = event.getActionMasked();
        if(isLastEventInGesture(actionMasked)) mDropGesture.cancel();
        else mDropGesture.submit();
        // Determine the hotbar slot
        float x = event.getX();
        // Ignore positions equal to mWidth because they would translate into an out-of-bounds hotbar index
        if(x < 0 || x >= mWidth) {
            // If out of bounds, cancel the hotbar gesture to avoid dropping items on last hotbar slots
            mDropGesture.cancel();
            return true;
        }
        int hotbarIndex = (int)MathUtils.map(x, 0, mWidth, 0, HOTBAR_KEYS.length);
        // Check if the slot changed and we need to make a key press
        if(hotbarIndex == mLastIndex) {
            // Only check for doubletapping if the slot has not changed
            if (hasDoubleTapped && !AllStaticSettings.disableDoubleTap) CallbackBridge.sendKeyPress(LwjglGlfwKeycode.GLFW_KEY_F);
            return true;
        }
        mLastIndex = hotbarIndex;
        int hotbarKey = HOTBAR_KEYS[hotbarIndex];
        CallbackBridge.sendKeyPress(hotbarKey);
        // Cancel the event since we changed hotbar slots.
        mDropGesture.cancel();
        // Only resubmit the gesture only if it isn't the last event we will receive.
        if(!isLastEventInGesture(actionMasked)) mDropGesture.submit();
        return true;
    }
/**
 * このオブジェクトが「LastEventInGesture」状態であるかを判定します。
 */
    private boolean isLastEventInGesture(int actionMasked) {
        return actionMasked == MotionEvent.ACTION_UP || actionMasked == MotionEvent.ACTION_CANCEL;
    }
/**
 * 「mc Scale」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private int mcScale(int input) {
        return (int)((mGuiScale * input)/ AllStaticSettings.scaleFactor);
    }
/**
 * 「run」メソッド。
 * このクラスに定義された機能メソッドです。
 */
    @Override
    public void run() {
        if (getParent() == null) return;

        HotbarType hotbarType = HotbarUtils.getCurrentType();
        if (hotbarType == HotbarType.AUTO) {
            adaptiveReset();
        } else {
            manualReset(AllSettings.getHotbarWidth().getValue().getValue(), AllSettings.getHotbarHeight().getValue().getValue(), false);
        }
    }
/**
 * 「on Layout Change」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
        // We need to check whether dimensions match or not because here we are looking specifically for changes of dimensions
        // and Android keeps calling this without dimensions actually changing for some reason.
        if(v.equals(mParentView) && (left != oldLeft || right != oldRight || top != oldTop || bottom != oldBottom)) {
            // Need to post this, because it is not correct to resize the view
            // during a layout pass.
            post(this::adaptiveReset);
        }
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/mouse/HotbarView.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.mouse

import android.content.Context
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.ViewParent
import com.arata.anim.AnimPlayer
import com.arata.anim.animations.Animations
import com.arata.yukarilauncher.event.single.MCOptionChangeEvent
import com.arata.yukarilauncher.event.single.RefreshHotbarEvent
import com.arata.yukarilauncher.event.value.HotbarChangeEvent
import com.arata.yukarilauncher.feature.MCOptions
import com.arata.yukarilauncher.setting.AllSettings
import com.arata.yukarilauncher.setting.AllStaticSettings
import com.arata.yukarilauncher.ui.subassembly.hotbar.HotbarType
import com.arata.yukarilauncher.ui.subassembly.hotbar.HotbarUtils
import com.arata.yukarilauncher.ui.view.GrabListener
import com.arata.yukarilauncher.utils.LwjglGlfwKeycode
import com.arata.yukarilauncher.Tools
import com.arata.yukarilauncher.utils.MathUtils
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.lwjgl.glfw.CallbackBridge

/**
 * Minecraftホットバー操作用のタッチビュー。
 * ホットバーの各スロットをタップしてアイテム選択やドロップ操作を行います。
 */
class HotbarView : View, View.OnLayoutChangeListener, Runnable {
    companion object {
        private val HOTBAR_KEYS = intArrayOf(
            LwjglGlfwKeycode.GLFW_KEY_1.toInt(), LwjglGlfwKeycode.GLFW_KEY_2.toInt(), LwjglGlfwKeycode.GLFW_KEY_3.toInt(),
            LwjglGlfwKeycode.GLFW_KEY_4.toInt(), LwjglGlfwKeycode.GLFW_KEY_5.toInt(), LwjglGlfwKeycode.GLFW_KEY_6.toInt(),
            LwjglGlfwKeycode.GLFW_KEY_7.toInt(), LwjglGlfwKeycode.GLFW_KEY_8.toInt(), LwjglGlfwKeycode.GLFW_KEY_9.toInt()
        )
    }

    private val mDoubleTapDetector = TapDetector(2, TapDetector.DETECTION_METHOD_DOWN)
    private var mParentView: View? = null
    private val mDropGesture = DropGesture(Handler(Looper.getMainLooper()))
    private val mGrabListener = GrabListener { isGrabbing ->
        mLastIndex = -1
        mDropGesture.cancel()
    }

    private var mWidth = 0
    private var mLastIndex = -1
    private var mGuiScale = 0
    private val adjustAnimPlayer = AnimPlayer()

    constructor(context: Context) : super(context) { init() }
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) { init() }
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) { init() }
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) { init() }

    private fun init() {
        alpha = 0f
        setBackgroundColor(Color.parseColor("#80E64242"))
        adjustAnimPlayer.duration(800)
        adjustAnimPlayer.apply(AnimPlayer.Entry(this, Animations.FadeOut))
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        EventBus.getDefault().register(this)
        val parent = parent
        if (parent is View) {
            mParentView = parent
            mParentView!!.addOnLayoutChangeListener(this)
        }
        adaptiveReset()
        CallbackBridge.addGrabListener(mGrabListener)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        CallbackBridge.removeGrabListener(mGrabListener)
        EventBus.getDefault().unregister(this)
    }

    /** @param event ホットバー更新イベント */
    @Subscribe
    fun event(event: RefreshHotbarEvent) {
        post(this)
    }

    /** @param event MCオプション変更イベント */
    @Subscribe
    fun event(event: MCOptionChangeEvent) {
        mGuiScale = MCOptions.mcScale
        post(this)
    }

    /** @param event ホットバーサイズ変更イベント */
    @Subscribe
    fun event(event: HotbarChangeEvent) {
        manualReset(event.width, event.height, true)
    }

    /** @return MarginLayoutParamsを取得 */
    private fun getMarginLayoutParams(): ViewGroup.MarginLayoutParams {
        val layoutParams = layoutParams
        if (layoutParams !is ViewGroup.MarginLayoutParams)
            throw RuntimeException("Incorrect LayoutParams type, expected ViewGroup.MarginLayoutParams")
        return layoutParams
    }

    /** 画面サイズに合わせてホットバーの位置とサイズを自動調整します。 */
    private fun adaptiveReset() {
        val marginLayoutParams = getMarginLayoutParams()
        val height: Int
        marginLayoutParams.width = mcScale(180).also { mWidth = it }
        marginLayoutParams.height = mcScale(20).also { height = it }
        marginLayoutParams.leftMargin = CallbackBridge.physicalWidth / 2 - mWidth / 2
        marginLayoutParams.topMargin = CallbackBridge.physicalHeight - height
        layoutParams = marginLayoutParams
    }

    /**
     * 指定されたサイズでホットバーを再設定します。
     * @param width 幅
     * @param height 高さ
     * @param playAnim アニメーションを再生するかどうか
     */
    private fun manualReset(width: Int, height: Int, playAnim: Boolean) {
        val marginLayoutParams = getMarginLayoutParams()
        marginLayoutParams.width = width.also { mWidth = it }
        marginLayoutParams.height = height
        marginLayoutParams.leftMargin = Tools.currentDisplayMetrics.widthPixels / 2 - width / 2
        marginLayoutParams.topMargin = Tools.currentDisplayMetrics.heightPixels - height
        layoutParams = marginLayoutParams
        if (playAnim) adjustAnimPlayer.start()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!CallbackBridge.isGrabbing()) return false
        val hasDoubleTapped = mDoubleTapDetector.onTouchEvent(event)

        val actionMasked = event.actionMasked
        if (isLastEventInGesture(actionMasked)) mDropGesture.cancel()
        else mDropGesture.submit()

        val x = event.x
        if (x < 0 || x >= mWidth) {
            mDropGesture.cancel()
            return true
        }
        val hotbarIndex = MathUtils.map(x, 0f, mWidth.toFloat(), 0f, HOTBAR_KEYS.size.toFloat()).toInt()

        if (hotbarIndex == mLastIndex) {
            if (hasDoubleTapped && !AllStaticSettings.disableDoubleTap)
                CallbackBridge.sendKeyPress(LwjglGlfwKeycode.GLFW_KEY_F.toInt())
            return true
        }

        mLastIndex = hotbarIndex
        val hotbarKey = HOTBAR_KEYS[hotbarIndex]
        CallbackBridge.sendKeyPress(hotbarKey)
        mDropGesture.cancel()
        if (!isLastEventInGesture(actionMasked)) mDropGesture.submit()
        return true
    }

    /** @param actionMasked イベントのアクション */
    private fun isLastEventInGesture(actionMasked: Int): Boolean {
        return actionMasked == MotionEvent.ACTION_UP || actionMasked == MotionEvent.ACTION_CANCEL
    }

    /** @param input 入力値（MCスケール適用後を返す） */
    private fun mcScale(input: Int): Int {
        return (mGuiScale * input / AllStaticSettings.scaleFactor).toInt()
    }

    override fun run() {
        if (parent == null) return

        val hotbarType = HotbarUtils.getCurrentType()
        if (hotbarType == HotbarType.AUTO) {
            adaptiveReset()
        } else {
            manualReset(AllSettings.hotbarWidth.value.getValue(), AllSettings.hotbarHeight.value.getValue(), false)
        }
    }

    override fun onLayoutChange(v: View, left: Int, top: Int, right: Int, bottom: Int, oldLeft: Int, oldTop: Int, oldRight: Int, oldBottom: Int) {
        if (v == mParentView && (left != oldLeft || right != oldRight || top != oldTop || bottom != oldBottom)) {
            post { adaptiveReset() }
        }
    }
}